### PR TITLE
chore(build): Move more of the Dart build to Broccoli.

### PR DESCRIPTION
**Pull request by mprobst**
@IgorMinar FYI.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -577,7 +515,6 @@ gulp.task('test.unit.cjs', ['build.js.cjs'], function () {
     var relPath = path.relative(__dirname, event.path).replace(/\\/g, "/");
     gulp.src(relPath)
       .pipe(gulpPlugins.rename({extname: '.'+ 'js'}))
-      .pipe(util.insertSrcFolder(gulpPlugins, CONFIG.srcFolderInsertion.js))
```
Comment (alexeagle) [inline gulpfile.js]: interesting - this wasn't needed here?

File: tools/broccoli/trees/dart_tree.ts [hunk 2]
```diff
@@ -1,32 +1,103 @@
...
+                '<script src="packages/browser/dart.js" type="text/javascript"></script>';
+  return content.replace('$SCRIPTS$', scriptTags);
+}
+
+interface MultiCopyOptions {
+  srcPath: string, targetPatterns: string[], exclude?: string[],
+}
+
+/**
+ * A writer that copies an input file from an input path into (potentially many) output locations
```
Comment (alexeagle) [inline tools/broccoli/trees/dart_tree.ts]: yum, we need this for the js trees also, should extract a helper in this directory (this change or follow-up)

File: tools/broccoli/trees/dart_tree.ts [hunk 3]
```diff
@@ -1,32 +1,103 @@
...
+              fsx.mkdirsSync(folder);
+              var outputPath = path.join(folder, fileName);
+              fs.writeFileSync(outputPath, data);
+            });
+          });
+        });
+  }
+}
+
+function stripModulePrefx(relativePath: string): string {
```
Comment (alexeagle) [inline tools/broccoli/trees/dart_tree.ts]: spelling: prefix

File: tools/broccoli/trees/dart_tree.ts [hunk 4]
```diff
@@ -41,7 +112,75 @@ module.exports = function makeDartTree() {
...
+  var mdTree = stew.rename(modulesFunnel(['**/*.dart.md']),
+                           relativePath => relativePath.replace(/\.dart\.md$/, '.md'));
+  // Copy all assets, ignore .js. and .dart. (handled above).
+  var docs = modulesFunnel(['**/*.md', '**/*.png', '**/*.html', '**/*.css'],
+                           ['**/*.js.md', '**/*.dart.md']);
+  return mergeTrees([licenses, mdTree, docs]);
+}
+
+module.exports = function makeDartTree() {
+  var sourceTree = mergeTrees([getSourceTree(), getHtmlSourcesTree()]);
```
Comment (alexeagle) [inline tools/broccoli/trees/dart_tree.ts]: yeah this reads well with the extracted functions, should do similar in JS trees

File: tools/broccoli/trees/dart_tree.ts [hunk 5]
```diff
@@ -1,32 +1,103 @@
...
+/**
+ * A writer that copies an input file from an input path into (potentially many) output locations
+ * given by a glob pattern.
+ */
+class MultiCopy extends Writer {
+  constructor(private inputTree, private options: MultiCopyOptions) { super(); }
 
-  // Transpile to dart.
-  var dartTree = ts2dart.transpile(modulesTree);
+  write(readTree: (tree) => Promise<string>, destDir: string): Promise<any> {
```
Comment (alexeagle) [inline tools/broccoli/trees/dart_tree.ts]: note: I'm working on a change now which adds a typing for the Tree

**Comment by alexeagle** [comment]
yum deleted lines

File: gulpfile.js (hunk 1 above)
Comment (mprobst) [inline gulpfile.js]: For js, the srcFolderInsertion was always a no-op. AFAICT, this was just cargo-culted all around.

File: tools/broccoli/trees/dart_tree.ts (hunk 2 above)
Comment (mprobst) [inline tools/broccoli/trees/dart_tree.ts]: Extracted.

File: tools/broccoli/trees/dart_tree.ts (hunk 3 above)
Comment (mprobst) [inline tools/broccoli/trees/dart_tree.ts]: Done.

File: tools/broccoli/trees/dart_tree.ts (hunk 5 above)
Comment (mprobst) [inline tools/broccoli/trees/dart_tree.ts]: Nice.

**Pull request closed without merging**